### PR TITLE
Refactor Datalog representation

### PR DESCRIPTION
**Pull request by bdahse**
Introduce a new Datalog constructor `Model`. The idea is to represent facts as a minimal model to avoid frequent marshalling between models computed by the solver and arrays consumed by Flix.

Models are immutable.

### CODE REVIEW

File: main/src/library/Fixpoint/Interpreter.flix [hunk 1]
```diff
@@ -23,13 +23,15 @@ namespace Fixpoint {
...
+
+    def interpretWithDatabase(db: Database[v], stmt: RamStmt[v]): Database[v] & Impure with Order[v], Eq[v], ToString[v] =
         let _ = notifyPreInterpret(stmt);
-        let db = MutMap.empty();
         evalStmt(db, stmt);
         db
 
     def evalStmt(db: Database[v], stmt: RamStmt[v]): Unit & Impure with Order[v], Eq[v], ToString[v] = match stmt {
-        case RamStmt.Facts(fs) =>
+        case RamStmt.Facts(fs) => // TODO: consider removing this.
```
Comment (bdahse) [inline main/src/library/Fixpoint/Interpreter.flix:34]: With this change, Datalog programs should no longer contain many facts. A program with many facts is represented as `Join(Model, Datalog)` where `Model` contains the facts and `Datalog` contains the rules. So it should no longer be a performance problem to handle facts by compiling them to `project (...) into A` statements.

**Comment by bdahse** [comment]
Now the database is not private to the interpreter anymore, so in the future a Database abstraction that is shared between Solver (Flix interface) and Interpreter may be a good idea.

**Comment by magnus-madsen** [comment]
> Now the database is not private to the interpreter anymore, so in the future a Database abstraction that is shared between Solver (Flix interface) and Interpreter may be a good idea.

Can you expand on what you mean?

**Comment by bdahse** [comment]
> 
> Can you expand on what you mean?

Look at `projectIntoX` or `factsOf`. They are working directly on the database representation that used to be internal to the interpreter. This leads to open coding of projecting a tuple into a relation in both the Solver and Interpreter. They have to take care to respect lattice semantics for example: When I union two models, I have to care if a predicate is latticenal to make sure I union with `lub`.

**Comment by magnus-madsen** [comment]
This is marked as draft. Let me know when it is ready for review.

**Comment by bdahse** [comment]
> This is marked as draft. Let me know when it is ready for review.

You can take a look now.

File: main/src/library/Fixpoint/Solver.flix [hunk 2]
```diff
@@ -40,12 +40,19 @@ namespace Fixpoint {
...
+            simplifyStmt |>
+            queryStmt |>
+            lowerStmt;
+        let model = match d {
+            case Datalog(_) => compiler(d) |> interpret |> toModel
+            case Model(_) => d
+            case Join(Model(m), cs) =>
+                let db = Map.map(Map.toMut, m) |> Map.toMut;
+                compiler(cs) |> interpretWithDatabase(db) |> toModel
+            case _ => bug!("Datalog normalization bug")
```
Comment (magnus-madsen) [inline main/src/library/Fixpoint/Solver.flix:54]: Can this case happen?

**Comment by magnus-madsen** [comment]
Merged! Thanks :)

File: main/src/library/Fixpoint/Solver.flix (hunk 2 above)
Comment (bdahse) [reply main/src/library/Fixpoint/Solver.flix:54]: It shouldn't. The logic in `union` is supposed to guarantee the composition is either:
 - A `Datalog` value (if both inputs are `Datalog` values).
 - A `Model` (if both inputs are models)
 - A `Join(Model, Datalog)` (if a `Model` and `Datalog` value is being composed, regardless of the order of composition).

**Pull request merged**